### ISSUE TITLE
p2eRatio() method in ion class gives incorrect results at low temperatures

### ISSUE THREAD

**Issue by wtbarnes** [chianti-atomic/ChiantiPy#55]
For a temperature array between 1e4 and 1e7 K spaced evenly in log space, the `p2eRatio()` method in the `ion` class returns incorrect proton to electron density ratios, particularly for temperatures below 10^5, according to those answers returned by the equivalent IDL CHIANTI routine [`proton_dens`](http://www.chiantidatabase.org/doc/utilities.html#PROTON_DENS). Ratios for T>10^5 K seem to be in pretty good agreement (i.e. within ~.1%).

It's not clear to me exactly what is causing this issue, but it could have something to do with the repeated interpolations that are done in [`p2eRatio`](http://chiantipy.readthedocs.io/en/latest/_modules/ChiantiPy/core/Ion.html#ion.p2eRatio). The code snippet below (using Eq. 7 of [Young et al., 2003](http://adsabs.harvard.edu/cgi-bin/nph-bib_query?bibcode=2003ApJS..144..135Y&db_key=AST&high=3a757f7f7a15254)) seems to give the correct answers (when compared with the IDL output):

```Python
import numpy as np
import scipy.interpolate
import ChiantiPy.tools as ch_tools

ioneq = ch_tools.io.ioneqRead(ioneqname='chianti')
abundance = ch_tools.io.abundanceRead(abundancename='sun_photospheric_2007_grevesse.abund')

abundance = abundance['abundance'][abundance['abundance']>0]
denom = np.zeros(len(ioneq['ioneqTemperature']))

for i in range(len(abundance)):
    for z in range(1,i+2):
        denom += z*ioneq['ioneqAll'][i,z,:]*abundance[i]

p2eratio = abundance[0]*ioneq['ioneqAll'][0,1,:]/denom
temperature = np.logspace(4,7,100)
f_interp = scipy.interpolate.interp1d(np.log10(ioneq['ioneqTemperature']),p2eratio,kind='cubic')
p2eratio_interp = f_interp(np.log10(temperature))
``` 

The equivalent IDL command is:
```IDL
!abund_file = '/path/to/chianti/dbase/abundance/sun_photospheric_2007_grevesse.abund'
temperature =  findgen(100)*(7-4)/99+4
p2eratio = proton_dens(temperature) 
```

**Comment by kdere** [comment]
I have just received my IDL license but have not installed it - probably today.  If you want to keep looking at this, go ahead.  I have a high regard for Peter's programming.  I am now hoping to do the ionization cross-sections/rates by comparing to what I did in IDL

**Comment by wtbarnes** [comment]
Yes I was using the IDL code as the baseline. I will try to fix this soon. I think it should be a pretty simple patch.

**Comment by kdere** [comment]
OK, that is fine.  I just got my IDL license installed but have not set up everything yet.

**Comment by wtbarnes** [comment]
In the `p2eRatio` method, `self.EDensity` is set as the numerator in the expression used to calculate the proton to electron ratio ([link](https://github.com/chianti-atomic/ChiantiPy/blob/master/ChiantiPy/core/Ion.py#L945)). If `eDensity` is set when the `ion` object is instantiated, this is overridden. 

What is the reason for setting the EDensity attribute here? It seems that in the case where `eDensity` is not set, this results in `self.EDensity` being set incorrectly.

**Comment by kdere** [comment]
That bit of code where
`self.EDensity = edensity`
definitely does not look right.  self.EDensity should never be overwritten
Do you want to continue with this or me?  It is fine with me if you do it.

**Comment by wtbarnes** [comment]
I can fix this when I implement the proposed changes to the `p2eRatio` method.

**Comment by wtbarnes** [comment]
Closing as this is Fixed by #74
